io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -1026,4 +1042,60 @@ impl LvsLvol for Lvol {
         }
         Ok(name)
     }
+
+    /// Resize a replica. The resize can be expand or shrink, depending
+    /// upon if required size is more or less than current size of
+    /// the replica.
+    async fn resize_replica(mut self, resize_to: u64) -> Result<Lvol, Error> {
+        let (s, r) = pair::<ErrnoResult<*mut spdk_lvol>>();
+        let mut ctx = ResizeCbCtx {
+            lvol: self.as_inner_ptr(),
+            sender: cb_arg(s),
+            req_size: resize_to,
+        };
+
+        unsafe {
+            vbdev_lvol_resize(
+                self.as_inner_ptr(),
+                resize_to,
+                Some(lvol_resize_cb),
+                &mut ctx as *mut _ as *mut c_void,
+            );
+        }
+
+        let cb_ret = r.await.expect("lvol resize callback dropped");
+
+        if let Ok(_lv) = cb_ret {
+            info!("Resized {:?} successfully", self);
+            Ok(Lvol::from_inner_ptr(self.as_inner_ptr()))
+        } else {
+            let status = cb_ret.unwrap_err();
+            error!("Resize {:?} failed, errno {}", self, status);
+            Err(Error::RepResize {
+                source: status,
+                name: self.name(),
+            })
+        }
+    }
+}
+
+extern "C" fn lvol_resize_cb(cb_arg: *mut c_void, errno: i32) {
+    let ctx = cb_arg as *mut ResizeCbCtx;
+    let (lvol, req_size) =
+        unsafe { (Lvol::from_inner_ptr((*ctx).lvol), (*ctx).req_size) };
+    let sender = unsafe {
+        Box::from_raw(
+            (*ctx).sender as *mut oneshot::Sender<ErrnoResult<*mut spdk_lvol>>,
+        )
+    };
+
+    if errno == 0 {
+        // Make sure resize worked, and account for metadata while comparing
+        // i.e. the actual size will be a little more than requested.
+        assert!(lvol.size() >= req_size);
```

Comment:
hmm seems like we should return error rather than crashing?

Reply:
This kind of situation might have a potential for leading to data integrity issues, hence marked assert. What would be the most severe error that we can return back here? We'll have to overwrite errno.

Reply:
Yeah we can overwrite errno to let control plane know

Reply:
Will overwrite errno to `EAGAIN` here because it has to be retried, assuming failure.